packages/deployer/subtasks/validate-storage.js

```diff
@@ -0,0 +1,75 @@
+const { subtask } = require('hardhat/config');
+const logger = require('@synthetixio/core-js/utils/logger');
+
+const { getSlotAddresses, findDuplicateSlots } = require('../internal/ast-helper');
+const { SUBTASK_VALIDATE_STORAGE, SUBTASK_CANCEL_DEPLOYMENT } = require('../task-names');
+
+subtask(SUBTASK_VALIDATE_STORAGE).setAction(async (_, hre) => {
+  logger.subtitle('Validating Storage usage');
+
+  const { contracts } = await _getSourcesAST(hre);
+
+  let errorsFound;
+  errorsFound = _findDuplicateStorageNamespaces(contracts) || errorsFound;
+  errorsFound = _findRegularStorageSlots() || errorsFound;
+  errorsFound = _findInvalidMutationsOnNamespaces() || errorsFound;
+
+  if (errorsFound) {
+    logger.error('Storate usage is not valid');
+    return await hre.run(SUBTASK_CANCEL_DEPLOYMENT);
+  }
+  logger.checked('Namespaces are valid');
+});
+
+async function _getSourcesAST(hre) {
```

Comment:
Wondering if this would have a better home in `internal/ast-helper`

Reply:
:) done